sourcegraph/cody
# Autocomplete: dynamically switch to multliline completions

## Context

- **TLDR**: we now use the completion's first line to determine if it starts the new multiline syntax node. If it does, we continue streaming until the completion is truncated or we reach the token sample limit.
- Closes https://github.com/sourcegraph/cody/issues/1619
- Adds a **feature flag** `cody-autocomplete-dynamic-multline-completions` to run an experiment on the updated logic.
- Adds an **experimental setting**, `cody.autocomplete.experimental.dynamicMultilineCompletions` to test the feature locally.
- Adds `completionPostProcessLogger` used for local debugging. It's disabled by default, and I will replace it with OpenTelemetry spans in a follow-up PR. I put a little effort into improving this logger because it's a throw-away code, but I need it here because debugging streaming-truncation bugs is much faster with it.
- Revamps the multiline truncation logic:
    - We no longer look for the next new sibling after the completion. This approach is unreliable because the parse tree often breaks because of incomplete code in the completion insert text.
    - Instead, we take the descendant at the current cursor position (or multiline trigger position), look for the farthest ancestor that starts at the same line, and then wait for a moment when completion generates more line than this syntax node has. This means we do not care about broken parse-tree after the node under the cursor:

```ts
// Start completion
function test() {
  █
}

// ---------------------------

// Continue streaming
function test() {
  █if (true) {
    console.log('hello world'
}

// ---------------------------

// Stop, because characters were added outside of `function test` syntax node boundaries
function test() {
  █if (true) {
    console.log('hello world'
}

function 
```
- ~~Need to merge https://github.com/sourcegraph/cody/pull/1891 first to reduce the diff in this PR.~~

## Screenshots

![Group 2](https://github.com/sourcegraph/cody/assets/3846380/79cb9e77-7230-4199-9509-45e2218305bf)
![Group 3](https://github.com/sourcegraph/cody/assets/3846380/f59f9bc0-689c-4b79-bfa9-a5151b77e926)
![Group 4](https://github.com/sourcegraph/cody/assets/3846380/b3c5185c-38a5-4cbd-b9e1-0b449acf91fd)
![Group 1](https://github.com/sourcegraph/cody/assets/3846380/11fe193f-0ccc-427b-bb0b-6fcfe0c30d4d)


## Follow-ups

- https://github.com/sourcegraph/cody/issues/1923
- https://github.com/sourcegraph/cody/issues/1924
- https://github.com/sourcegraph/cody/issues/1920
- https://github.com/sourcegraph/cody/issues/1921

## Merged preparation PRs

- https://github.com/sourcegraph/cody/pull/1709
- https://github.com/sourcegraph/cody/pull/1784
- https://github.com/sourcegraph/cody/pull/1888
- https://github.com/sourcegraph/cody/pull/1891
- https://github.com/sourcegraph/cody/pull/1893

## Test plan

- Tested manually:
    - Ensure that code completions are generated without changes on this branch.
    - Ensure that code completions are dynamically switched to multiline based on the first line when `cody.autocomplete.experimental.dynamicMultilineCompletions` is set to true.
- Existing tests pass.
- The functionality is disabled by default. I'm going to add more unit tests for the new logic in a separate PR (see planned follow-ups section).
